### PR TITLE
VEN-1 | Parse single file for function declarations

### PR DESCRIPTION
**Pull request by bokimilinkovic**
Purpose of this PR is to create functionality to crawl all public methods of a single file and write in a simple yaml output their declarations... It will be used furthermore for incoming features

### CODE REVIEW

File: test-crawler/extractor/extractor.go [hunk 1]
```diff
@@ -103,17 +128,17 @@ func parseContent(content string, treeCursor *sitter.TreeCursor, filePath string
 	var annotationParser a.Parser
 
 	fileData.Metadata = getMetadata(content, treeCursor, &annotationParser)
-	functions := getFunctionNodes(content, treeCursor, &annotationParser)
+	funcNodes := getFunctionNodes(content, treeCursor, &annotationParser)
 
-	for _, function := range functions {
+	for _, function := range funcNodes {
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:133]: Rename `function` to `funcNode`. See line 141: 
`function.Function.Name` would then be `funcNode.Function.Name`. 

Which is much clearer, no?

File: test-crawler/extractor/extractor.go [hunk 2]
```diff
@@ -163,9 +188,12 @@ func getMetadata(content string, treeCursor *sitter.TreeCursor, parser *a.Parser
 	return &meta
 }
 
+// this method extracts functions and methods.
+// Function can have 2 types: function_declaration (for example contructor)
+// and method_declaration (can be exported and unexported)
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:193]: Please format this comment in line with GoDoc standards, and add an explanation of the return value.

File: test-crawler/extractor/extractor.go [hunk 3]
```diff
@@ -204,17 +232,41 @@ func getFunctionNodes(content string, treeCursor *sitter.TreeCursor, parser *a.P
 				}
 
 				funcAnnoPair = append(funcAnnoPair, struct {
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:234]: We are redefining this struct here and in the function return values declaration.
Therefore, it is a good idea to declare the struct outside the function: we will avoid code duplication, and make the code more readable.

File: test-crawler/extractor/extractor.go [hunk 4]
```diff
@@ -204,17 +232,41 @@ func getFunctionNodes(content string, treeCursor *sitter.TreeCursor, parser *a.P
...
+			if child.Type() == string(METHOD_DECLARATION) {
+				funcName = content[child.Child(2).StartByte():child.Child(2).EndByte()]
+				public := unicode.IsUpper(rune(funcName[0]))
+				params := ""
+				returnValues := ""
+				if child.Child(1).Type() == string(PARAMETER_LIST) {
+					params = content[child.Child(3).StartByte():child.Child(3).EndByte()]
+					returnValues = content[child.Child(4).StartByte():child.Child(4).EndByte()]
+				}
+				funcAnnoPair = append(funcAnnoPair, struct {
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:257]: See [my previous comment](https://github.com/filecoin-project/system-test-matrix/pull/943/files#r1026287192)

File: test-crawler/extractor/extractor.go [hunk 5]
```diff
@@ -204,17 +232,41 @@ func getFunctionNodes(content string, treeCursor *sitter.TreeCursor, parser *a.P
...
+						Name: funcName,
+					},
 				})
 
 				node = nil
 				funcName = ""
 				isIgnored = false
 			}
+			if child.Type() == string(METHOD_DECLARATION) {
+				funcName = content[child.Child(2).StartByte():child.Child(2).EndByte()]
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:249]: Add comment(s) explaining what the Child(2,3,4...) are.

File: test-crawler/extractor/extractor.go [hunk 6]
```diff
@@ -301,3 +353,15 @@ func makeID(filePath string, funcName string, behavior string) string {
 	hash := md5.Sum([]byte(fmt.Sprintf("%s_%s_%s", filePath, funcName, behavior)))
 	return string(hex.EncodeToString(hash[:]))
 }
+
+func parseContentForPublicFunctions(content string, cursor *sitter.TreeCursor) ([]c.FunctionAnnotation, error) {
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:357]: The method is not named properly (it does not return only public functions, but all functions), and it does not have a GoDoc comment.

Comment (bokimilinkovic) [reply test-crawler/extractor/extractor.go:357]: I think that unexported functions does not need comments by default for GoDoc, but I will add simple one explaining what that functions does.

Comment (StefanGajic) [reply test-crawler/extractor/extractor.go]: As far as I know there will be linter error for functions that are exported and have no comment. But a good practice would be for all the functions both exported and unexported to have a simple comment.

File: test-crawler/extractor/extractor.go (hunk 5 above)
Comment (brdji) [reply test-crawler/extractor/extractor.go:266]: A bit too long, but good. I would remove this:
```
First iteration through children nodes:
Child.Type = method_declaration
Child.Value = whole function body func(*x)(a,b)c{...}
```
Since we are not handling the type/value in this part of the function. Also, I would remove the "has a type XX, and represents..." text, because it is superfluous.

File: test-crawler/extractor/extractor.go [hunk 7]
```diff
@@ -203,18 +234,53 @@ func getFunctionNodes(content string, treeCursor *sitter.TreeCursor, parser *a.P
...
+				Child.Value = whole function body func(*x)(a,b)c{...}
+
+				child.Child(1) is a first part of function declaration, and has type of parameter_list: (e *ProofEventStream)
+				child.Child(2): has type of field_identifier  and represents function's name: ListenProofEvent
+				child.Child(3): has type parameter_list and represents input parameters:  (ctx context.Context, policy *types2.ProofRegisterPolicy)
+				child.Child(4): has type parameter_list and represent return parameters:  (<-chan *types2.RequestEvent, error)
+			*/
+			if child.Type() == string(METHOD_DECLARATION) {
+				funcName = content[child.Child(2).StartByte():child.Child(2).EndByte()]
+				public := unicode.IsUpper(rune(funcName[0]))
```
Comment (brdji) [inline test-crawler/extractor/extractor.go:267]: It would be a good idea to extract this to a separate function, e.g. `function isPublic(string functionName)`

File: test-crawler/main.go [hunk 8]
```diff
@@ -26,6 +26,11 @@ func main() {
 
 	config := NewConfig()
 
+	// crawlRepoBehaviorsAndSaveToJSON(config)
```
Comment (brdji) [inline test-crawler/main.go:29]: Leftover code? This should not be committed.

File: test-crawler/config.go [hunk 9]
```diff
@@ -21,6 +21,7 @@ type Config struct {
 	Language   string     `yaml:"lang_mode"`
 	IndentJSON bool       `yaml:"indent_json"`
 	Ignore     []string   `yaml:"ignore"`
+	SingleFile string     `yaml:"single_file"`
```
Comment (brdji) [inline test-crawler/config.go:24]: There is no use for this in the config. If you wanted to test a single file, you could simply hardcode a string in the `crawlSingleFileForMethods` method.

Instead of a single string, you have:
1. Added a line to the config files
2. Read the line from the config

Which is pointless, because the crawler will never read from the config a single file to crawl.
tl;dr remove the `single file` config line from all usages.

File: test-crawler/extractor/extractor.go (hunk 6 above)
Comment (brdji) [reply test-crawler/extractor/extractor.go]: > As far as I know there will be linter error for functions that are exported and have no comment. But a good practice would be for all the functions both exported and unexported to have a simple comment.

Correct. I would prefer to have a simple comment for each function we add, even if the function is unexported.

File: test-crawler/main.go [hunk 10]
```diff
@@ -26,6 +26,11 @@ func main() {
 
 	config := NewConfig()
 
+	crawlRepoBehaviorsAndSaveToJSON(config)
+	crawlSingleFileForMethods("../repo-to-crawl/system/subsystem/file.go") // // ex: ../repo-to-crawl/venus-gateway/proofevent/proof_event.go
```
Comment (brdji) [inline test-crawler/main.go:30]: Leftover code - should not be committed.

**Comment by brdji** [comment]
The name `main_test.go` is incorrect. It does not test the `main` module, it tests the behavior of the function crawler. Please rename it to a more fitting name (e.g. `crawler_test.go` is fine for now)

**Review by brdji — APPROVED** [review]
